core/src/main/scala/kafka/log/Log.scala

```diff
@@ -2686,3 +2670,50 @@ object LogMetricNames {
     List(NumLogSegments, LogStartOffset, LogEndOffset, Size)
   }
 }
+
+sealed trait SegmentDeletionReason {
+  def reasonString(log: Log, toDelete: Iterable[LogSegment]): String
+}
+
+case object RetentionMsBreachDeletion extends SegmentDeletionReason {
```

Comment:
nit: is it necessary to add `Deletion` to all of these? Maybe only `LogDeletion` needs it since it is referring to deletion of the log itself.